### PR TITLE
Google Speech Command Data Loader

### PR DESCRIPTION
**Pull request by ljj7975**
main changes:
- `GSCDatasetPreprocessor`: reorganize filenames train/dev/test set from the dataset
- `GSCDataset`: supports file loading and adding noise for train set
- `AudioDataLoader`: get each audio sample and preprocess it using MFCC/PCEN

note that:
- the main focus was the readability. Therefore, I iterate over the dataset twice (once to organize by class, once to reorganize by dataset).
- separated out `AudioDataLoader` in case we add different dataset in the future (Crawled data)
- Caching is not supported yet
- Loading a class using `get_attr` will be updated in the future PR

### CODE REVIEW

File: data_loader/audio_data_loader.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from torch.utils.data import DataLoader
+
+class AudioDataLoader(DataLoader):
+    def __init__(self, data_loader_config, dataset_config):
+        self.dataset = GSCDataset(dataset_config)
+
+        self.audio_preprocessing = data_loader_config["audio_preprocessing"]
+        self.audio_processor = AudioProcessor()
+
+        super(AudioDataLoader, self).__init__(
```
Comment (daemon) [inline data_loader/audio_data_loader.py]: `super()` is sufficient in Python 3.

File: data_loader/gsc_dataset.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                class_name = unknown_class_name
+
+            if class_name not in audio_files_by_class:
+                audio_files_by_class[class_name] = []
+                audio_counts_by_class[class_name] = 0
+
+            count = 0
+            for file_name in os.listdir(dir_path):
+
+                if "wav" not in file_name:
```
Comment (daemon) [inline data_loader/gsc_dataset.py]: I would look at `pathlib.Path.suffix` for determining file type.

File: data_loader/gsc_dataset.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            label = config["target_class"].index(class_name)
+            self.label_mapping[label] = class_name
+
+            for audio_file in audio_list:
+                bucket = self.get_bucket_from_file_name(audio_file, config["group_speakers_by_id"])
+                self.distribute_to_dataset(bucket, audio_file, label)
+
+        # unknwon class
+        if config["unknown_class"]:
+            unknwon_label = len(config["target_class"])
```
Comment (daemon) [inline data_loader/gsc_dataset.py]: unknwon -> unknown

File: main.py [hunk 4]
```diff
@@ -116,22 +150,22 @@ def main(mode, config):
...
+            train_log["loss"].append(total_loss / len(train_data_loader))
+            train_log["metric"].append(total_metric / len(train_data_loader))
 
             if epoch % config["checkpoint_frequency"] == 0:
                 torch.save(model.state_dict(), checkpoint_path_template.format(epoch))
 
                 print(f"epochs {epoch}")
-                print(f"\tloss: {training_log['loss'][-1]}")
-                print(f"\tmetric: {training_log['metric'][-1]}")
+                print(f"\tloss: {train_log['loss'][-1]}")
```
Comment (daemon) [inline main.py]: String single vs double quotation convention should be consistent within a project.

**Review by daemon — APPROVED** [review]

File: data_loader/gsc_dataset.py (hunk 3 above)
Comment (ljj7975) [reply data_loader/gsc_dataset.py]: lool thanks. good catch

File: main.py (hunk 4 above)
Comment (ljj7975) [reply main.py]: ya, right, I should keep it consistent
